src/platform_impl/windows/event_loop.rs

```diff
@@ -343,40 +350,65 @@ impl EventLoop {
         status
     }
 
-    /// Wait for one message and dispatch it, optionally with a timeout
-    fn wait_and_dispatch_message(&mut self, timeout: Option<Duration>) {
-        fn get_msg_with_timeout(msg: &mut MSG, timeout: Option<Duration>) -> PumpStatus {
+    /// Waits until new event messages arrive to be peeked.
+    /// Doesn't peek messages itself.
+    ///
+    /// Parameter timeout is optional. This method would wait for the smaller timeout
+    /// between the argument and a timeout from control flow.
+    fn wait_for_messages(&mut self, mut timeout: Option<Duration>) {
+        // Set upper limit for waiting time to avoid overflows.
+        // I chose 50 days as a limit because it is used in dur2timeout.
+        const FIFTY_DAYS: Duration = Duration::from_secs(50_u64 * 24 * 60 * 60);
+        // Waitable timers use 100 ns intervals to indicate due time.
+        // https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-setwaitabletimer#parameters
+        // And there is no point waiting using other ways for such small timings
+        // because they are even less precise (can overshoot by few ms).
+        const MIN_WAIT: Duration = Duration::from_nanos(100);
+
+        fn create_high_precision_timer() -> Option<HANDLE> {
             unsafe {
-                // A timeout of None means wait indefinitely (so we don't need to call SetTimer)
-                let timer_id = timeout.map(|timeout| SetTimer(0, 0, dur2timeout(timeout), None));
-                let get_status = GetMessageW(msg, 0, 0, 0);
-                if let Some(timer_id) = timer_id {
-                    KillTimer(0, timer_id);
-                }
-                // A return value of 0 implies `WM_QUIT`
-                if get_status == 0 {
-                    PumpStatus::Exit(0)
+                let handle: HANDLE = CreateWaitableTimerExW(
+                    ptr::null(),
+                    ptr::null(),
+                    CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+                    TIMER_ALL_ACCESS,
+                );
+                // CREATE_WAITABLE_TIMER_HIGH_RESOLUTION is supported only after
+                // Win10 1803 but it is already default option for rustc
+                // (std uses it to implement `std::thread::sleep`).
+                if handle == 0 {
+                    None
                 } else {
-                    PumpStatus::Continue
+                    Some(handle)
                 }
             }
         }
```

Comment:
I would rather make these freestanding functions than having them inlined into `wait_for_messages`.

Reply:
OK, would fix tomorrow. I just kept the structure that was used before.